File: algs/graph/weighted/prim.py

#  Skiena 2008

# Prim(G)
#   Select an arbitrary vertex `s` to start tree from
#   While (there exists nontree vertices)
#       Select edge of minimum weight betwee a tree and nontree vertex
#       Add the selected edge and vertex to the tree Tprim

from collections import defaultdict

def prim(graph, start):
    parent = defaultdict(lambda: None)
    distance = defaultdict(lambda: float('inf'))
    tree = set()
    v = start 

    while v not in tree:
        tree.add(v)

        # update each of v's edges
        for edge in graph.edges(v):
            weight = edge.weight
            w = edge.vertex
            if (w not in tree) and (weight < distance[w]):
                distance[w] = weight
                parent[w] = v

        # select the vertex with min edge weight to MST
        v = min([v for v in graph.vertices if v not in tree], key=lambda v: distance[v])



    

File: algs/graph/weighted/prim.rb
# Prim's algorithm

require 'set'


EdgeNode = Struct.new(:weight, :vertex)

def prim(graph, start)

  parent = Hash.new {-1}
  distance = Hash.new {FLOAT::INFINITY}
  tree = Set.new


  v = start
  until v.nil?
    tree.add(v)

    graph.edges(v).each do |edge|
      w = edge.vertex
      if (not tree.member? w) and (edge.weight < distance[w])
        distance[w] = edge.weight
        parent[w] = v
      end
    end

    v = graph.vertices.select{|v| not tree.member? v }.min_by {|v| distance[v] }

  end
end
